### PR TITLE
Minimal api server

### PR DESCRIPTION
**Pull request by pzhokhov**
code for the API server and tests. The tests talk to both API server hosted on google app engine (our favourite https://studio-ed756.appspot.com), as well as local (but the latter are skipped in jenkins because they require GC or AWS credentials).

### CODE REVIEW

File: studio/debugtools.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import __builtin__
```
Comment (pzhokhov) [inline studio/debugtools.py]: this file is not actually being used currently, but it came it handy when I was debugging file descriptor leaks, so I decided to keep it here.

File: studio/gcloud_artifact_store.py [hunk 2]
```diff
@@ -16,42 +17,46 @@ def __init__(self, config, verbose=10, measure_timestamp_diff=True):
...
-            self.bucket = self.client.get_bucket(config['bucket'])
+            self.bucket = self.getclient().get_bucket(config['bucket'])
         except BaseException as e:
             self.logger.exception(e)
-            self.bucket = self.client.create_bucket(config['bucket'])
+            self.bucket = self.getclient().create_bucket(config['bucket'])
 
         super(GCloudArtifactStore, self).__init__(measure_timestamp_diff)
 
+    def getclient(self):
```
Comment (pzhokhov) [inline studio/gcloud_artifact_store.py]: this is a tricky one - apparently, credentials that are created when one initializes storage.Client() expire after a while, so same instance of storage.Client() cannot be used for too long. Was causing problems within google app engine.

File: studio/http_artifact_store.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
+import logging
+import requests
+
+from tartifact_store import TartifactStore
+logging.basicConfig()
+
+
+class HTTPArtifactStore(TartifactStore):
```
Comment (pzhokhov) [inline studio/http_artifact_store.py]: unlike other *ArtifactStore classes, this class only communicates to one artifact with a specific (signed) url, so it has to be instantiated for every artifact.

File: studio/http_provider.py [hunk 4]
```diff
@@ -145,13 +154,7 @@ def checkpoint_experiment(self, experiment):
         self._raise_detailed_error(request)
         artifacts = request.json()['artifacts']
 
-        for tag, art in experiment.artifacts.iteritems():
-            if 'local' in art.keys():
-                art['key'] = artifacts[tag]['key']
```
Comment (karlmutch) [inline studio/http_provider.py]: I could not find these keys added to the s3 storage implementation code. Are these keys still generated for all storage types?

Comment (pzhokhov) [reply studio/http_provider.py]: I am not sure I understand the question... This logic is specific to the db provider that talks to api server; and it simply uploads the file using the provided url. Does that somehow break the behavior that you expect?

Comment (karlmutch) [reply studio/http_provider.py]: Thanks, answered your your comment in that this is for the DB provider not for the JSON queue messages.  Thanks

**Pull request merged**